### PR TITLE
Use Bash features instead of `dirname` and `basename`

### PR DESCRIPTION
**Pull request by gaelicWizard**
## Description
Reduce subshells and external commands by using _Bash_'s string processing features instead of `dirname` and `basename`.

## Motivation and Context
This is part of a set of patches to reduce usage of subshells and external binaries in _Bash It_.

## How Has This Been Tested?
Tested locally.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code follows the code style of this project.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] If I have added a new file, I also added it to ``clean_files.txt`` and formatted it using ``lint_clean_files.sh``.
- [x] I have added tests to cover my changes, and all the new and existing tests pass.

### CODE REVIEW

File: install.sh [hunk 1]
```diff
@@ -184,7 +184,7 @@ if [[ $no_modify_config ]] && [[ $append_to_config ]]; then
 	exit 1
 fi
 
-BASH_IT="$(cd "$(dirname "$0")" && pwd)"
+BASH_IT="${BASH_SOURCE%/*}"
```
Comment (NoahGorny) [inline install.sh:187]: this gives us the relative path of bash-it and not the absolute one. it works for the installation, but lets at least add a comment about it here (or change to absolute path like before)

**Review by NoahGorny — CHANGES_REQUESTED** [review]
Added a small nit, otherwise looks great :smile:

File: install.sh (hunk 1 above)
Comment (gaelicWizard) [reply install.sh:187]: I think add a comment to the template `.bash_profile`, sound good?

Comment (gaelicWizard) [reply install.sh:187]: Except we modify the template during installation by replacing `{{BASH_IT}}`, so it makes more sense to make sure this one is absolute. ok

**Review by NoahGorny — APPROVED** [review]
Looks great!
I will let others chime in if they want :smile:

**Comment by davidpfarrell** [comment]
Hi All!

In theory, the idea is nice - you avoid the subshell (but lose _some_ explicitness)

In practice, the two approaches are not _completely_ identical.  For a short discussion on the matter:
* https://stackoverflow.com/questions/22401091/bash-variable-substitution-vs-dirname-and-basename

_stolen and reformatted table of mismatches_
```
file        basename $file  ${file##*/}  dirname $file  ${file%/*}
/aaa/bbb/   bbb             ''           /aaa           /aaa/bbb  
/           /               ''           /              ''        
/aaa        aaa             aaa          /              ''        
aaa         aaa             aaa          .              aaa       
aaa/        aaa             ''           .              aaa       
aaa//       aaa             ''           .              aaa/      

note: '' => empty string (quotes not really generated)
```

We'd have to be **very** sure that the format of the variable we were using didn't match any of these patterns, ie. Don't compute _basename_ from a directory with tailing `/` or don't compute _dirname_ on a directory off of root.

I think we might be better off just biting the bullet on the subshells for `dirname` and `basename`.

**Comment by NoahGorny** [comment]
@davidpfarrell - great feedback!
Very interesting. After seeing this difference I also agree with David here. I am all in for avoiding subshell, but not at the cost of unexplained behaviours and possible bugs!

Let us know what you think @gaelicWizard !

**Comment by gaelicWizard** [comment]
Thanks for the link to examples! You may have noticed, my whole brainworm is about edge cases so I'm grateful for the specific examples. Give me a few to look through these 😃

**Comment by NoahGorny** [comment]
Hi @gaelicWizard, as discussed, I currently prefer to not merge this PR, but focus on the other (great!) PRs that you did. I still want to hear what you think about this- so I am currently waiting for you

**Comment by gaelicWizard** [comment]
My apologies for the radio silence! I had intended to dun the statistics linked in @davidpfarrell's comment but just haven't found time. Hopefully this weekend 😃

**Comment by gaelicWizard** [comment]
The table of behavior differences between `dirname`/`basename` and the shell version boils down to two cases:
1. trailing slash / no file name, and
2. no slash / no directory / not a path (just a file).

Therefore I've now split this PR in to three commits:
1. Uses of `dirname` which don't hit either edge case,
2. Uses of `basename` which don't hit either edge case, and
3. Uses of `dirname` which may hit the "blank" edge case.

In the first two commits, the parameter is specified with a full path within the immediately preceding lines so we simply never hit the case where a (possibly user-supplied) unexpected string is provided that may have too many or too few slashes.

The third commit ("dirname (2 of 2)") has cases of directory walking or accepting user-specified path specifiers. These may or may not hit the two identified edge cases and definitely need something more/other than just string truncation. (Unrelated to `dirname`, the directory walking bothers me, so I may look in to some alternative.)

Reducing subshells and external binaries is a valuable goal IMO; I normally expect my shell to start in under a couple of seconds. I've seen comments in other issues and PRs indicating that some folks have shells that take *tens of seconds* to start, which is just bananas to me. So, I'd like to submit the first two commits under a separate PR, which I hope will result in an uncontroversial merge, and leave the third commit out. (I'd like to close this PR with all three for reference.) Let me know if this sounds reasonable?

---

The linked [SO thread]( https://stackoverflow.com/questions/22401091/bash-variable-substitution-vs-dirname-and-basename ) clearly lays out the performance cost of subshells and external binaries:
```
real    0m33.455s
user    0m10.194s
sys     0m18.106s
command basename
===================================

real    0m0.246s
user    0m0.237s
sys     0m0.007s
varsub basename
===================================

real    0m30.562s
user    0m10.115s
sys     0m17.764s
command dirname
===================================

real    0m0.237s
user    0m0.226s
sys     0m0.007s
varsub dirname
```

Here's the same timings from my MBP:
```

real	0m26.495s
user	0m8.437s
sys	0m14.692s
command basename
===================================

real	0m0.163s
user	0m0.160s
sys	0m0.004s
varsub basename
===================================

real	0m26.436s
user	0m8.560s
sys	0m14.475s
command dirname
===================================

real	0m0.123s
user	0m0.118s
sys	0m0.003s
varsub dirname
```

Here's another illustration I like of the resource cost as well:
`$ echo "$(echo "$(echo "$(echo "$(pstree -p $$)")")")"` or `$ echo "$(echo "$(echo "$(echo "$(ps ww -s $$)")")")"` shows us:
```
$ echo "$(echo "$(echo "$(echo "$(pstree -p $$)")")")"
-+= 00001 root /sbin/launchd
 \-+= 30593 pellican /usr/bin/SCREEN -D -m -ln -fn -S gnu.screen
   \-+= 72426 root login -pflq pellican /bin/bash
     \-+= 72427 pellican bash
       \-+- 87344 pellican bash
         \-+- 87345 pellican bash
           \-+- 87346 pellican bash
             \-+- 87347 pellican bash
               \-+- 87348 pellican pstree -p 72427
                 \--- 87349 root ps -axwwo user,pid,ppid,pgid,command
```

**Pull request closed without merging**